aten/src/ATen/native/sparse/SparseCsrTensor.cpp

```diff
@@ -8,12 +8,84 @@
 #include <ATen/SparseCsrTensorImpl.h>
 #include <ATen/SparseCsrTensorUtils.h>
 #include <ATen/SparseTensorImpl.h>
+#include <ATen/InitialTensorOptions.h>
 
 namespace at {
 namespace native {
 
 using namespace at::sparse_csr;
 
+namespace {
+
+
+} // end anonymous namespace
+
+void _validate_sparse_csr_tensor_args(const Tensor& crow_indices, const Tensor& col_indices, const Tensor& values, IntArrayRef size) {
+  // Layout Invariants
+  TORCH_CHECK(
+      col_indices.layout() == kStrided && col_indices.is_contiguous(),
+      "expected col_indices to be a strided and contiguous tensor");
+
+  TORCH_CHECK(
+      crow_indices.layout() == kStrided && crow_indices.is_contiguous(),
+      "expected crow_indices to be a strided and contiguous tensor");
+
+  TORCH_CHECK(
+      values.layout() == kStrided && values.is_contiguous(),
+      "expected values to be a strided and contiguous tensor");
+
+  // Shape and Strides invariants
+  TORCH_CHECK(
+      size.size() == 2,
+      "size of a CSR tensor must be of length 2, but got: ",
+      size.size());
+  TORCH_CHECK(
+      crow_indices.dim() == 1,
+      "crow_indices must have dim=1 but got crow_indices.dim()=",
+      crow_indices.dim());
+  TORCH_CHECK(
+      col_indices.dim() == 1,
+      "col_indices must have dim=1 but got col_indices.dim()=",
+      col_indices.dim());
+  TORCH_CHECK(
+      values.dim() == 1,
+      "values must have dim=1 but got values.dim()=",
+      values.dim());
+  // Note, this check also enforces `crow_indices.numel() >= 1`
+  TORCH_CHECK(
+      crow_indices.numel() == (size[0] + 1),
+      "crow_indices.numel() must be size(0) + 1, but got: ",
+      crow_indices.numel());
+  TORCH_CHECK(
+      col_indices.size(0) == values.size(0),
+      "col_indices and values must have equal sizes, but got col_indices.size(0): ",
+      col_indices.size(0),
+      ", values.size(0): ",
+      values.size(0));
+
+  // Indices invariants
+  AT_DISPATCH_INDEX_TYPES(crow_indices.scalar_type(), "csr_construct_check", [&] {
+    Tensor crow_indices_cpu = crow_indices.to(kCPU);
```

Comment:
OK, so we're back to the slower version? (I was reasonably convinced by @pearu's comment that it is faster to only transfer the indices you needed, which I agree is true)

Reply:
Previous to this update, only access to first and last element was required, however in order to completely cover the CSR invariants listed [here](https://github.com/pearu/pearu.github.io/blob/pearu-csr_invariants/csr_tensor_invariants.md):
> `5.3 crow_indices[i-1] <= crow_indices[i] for all i=1,...,nrows`

Access to the whole array is needed. We can go back to the  faster version only if this condition is skipped. cc @pearu

Reply:
I think we should keep testing 5.3 because:
- when the invariant 5.3 is not satisfied and we don't catch it, python will very likely crash when performing operations with CSR tensor with invalid `crow_indices`
- there is a way to skip expensive testing of invariants